### PR TITLE
Add cognitive complexity rule

### LINKED ISSUE #162: Rule: Cognitive Complexity of methods should not be too high
Implements [RSPEC-3776](https://jira.sonarsource.com/browse/RSPEC-3776)

### PR DESCRIPTION
**Pull request by ghost**
Fixes #162 

I executed current implementation against ruling to see repartition of cognitive complexity by function/method, with the following results. From my point of view, 15 seems to be a good starting value.

criteria | total | ratio
-- | -- | --
">0" | 32,227 | 100.00%
">5" | 9,153 | 28.40%
">10" | 4,641 | 14.40%
">15" | 2,897 | 8.99%
">20" | 1,885 | 5.85%
">50" | 367 | 1.14%

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,6 +49,7 @@ How does it work?
 * The output of functions that don't return anything should not be used ([`no-use-of-empty-return-value`]) ([`requires type-check`])
 * Variables should be declared before they are used ([`no-variable-usage-before-declaration`]) ([`requires type-check`])
 * Type aliases should be used ([`use-type-alias`]) ([`requires type-check`])
+* Cognitive Complexity of functions should not be too high ([`cognitive-complexity`])
```
Comment (vilchik-elena) [inline README.md]: @m-g-sonar we sort rules alphabetically by rule key

File: README.md [hunk 2]
```diff
@@ -84,6 +85,7 @@ How does it work?
 [`no-variable-usage-before-declaration`]: ./sonarts-core/docs/rules/no-variable-usage-before-declaration.md
 [`use-type-alias`]: ./sonarts-core/docs/rules/use-type-alias.md
 [`requires type-check`]: https://palantir.github.io/tslint/usage/type-checking/
+[`cognitive-complexity`]: ./sonarts-core/docs/rules/cognitive-complexity.md
```
Comment (vilchik-elena) [inline README.md]: @m-g-sonar same here

File: sonarts-core/tests/rules/cognitiveComplexityRule/cognitiveComplexityRule.lint.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
+// tslint:disable
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/cognitiveComplexityRule/cognitiveComplexityRule.lint.ts]: @m-g-sonar you can drop this line, it's leftover

File: sonarts-core/tests/runRule.ts [hunk 4]
```diff
@@ -51,6 +51,24 @@ export interface RuleRunResult {
...
+  return checkRule(Rule, lintFileName, ruleArguments);
+}
+
+/**
+ * Run rule againts a specific lint file from a test directory
+ * Use from the test: runRule(Rule, "myTestDir", "myTestFile"), ommitting '.lint.ts' suffix on test file
+ */
+export function runRuleWithLintFile(
+  Rule: any,
+  testDirName: string,
```
Comment (vilchik-elena) [inline sonarts-core/tests/runRule.ts]: @m-g-sonar it's dir here, and ruleName in deeper call. Let's keep one abstraction

File: sonarts-core/docs/rules/cognitive-complexity.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# cognitive-complexity
+
+Cognitive Complexity is a measure of how hard the control flow of a function is to understand. Functions with high Cognitive Complexity will be difficult to maintain.
+
+## See
+
+* [Cognitive Complexity](http://redirect.sonarsource.com/doc/cognitive-complexity.html)
```
Comment (vilchik-elena) [inline sonarts-core/docs/rules/cognitive-complexity.md]: @m-g-sonar add configuration section (like I did for mccabe-complexity)

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+  threshold: number;
+
+  constructor(sourceFile: ts.SourceFile, options: tslint.IOptions, threshold: number) {
+    super(sourceFile, options);
+    this.threshold = threshold;
+  }
+
+  public visitNode(node: ts.Node): void {
+    if (is(node, ...Rule.TARGETED_KINDS)) {
+      const body = (node as ts.FunctionLikeDeclarationBase).body as ts.Node;
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar what for this cast ` as ts.Node`? I think you can just drop it

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+      if (body) {
+        const complexity = this.getComplexity(body);
+        if (complexity > this.threshold) {
+          this.addFailureAtNode(
+            FunctionWalker.reportNode(node),
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar why didn't you use `functionLikeMainToken` from navigation?

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+      case ts.SyntaxKind.GetAccessor:
+      case ts.SyntaxKind.SetAccessor:
+        return "accessor";
+      default:
+        return "function";
+    }
+  }
+}
+
+class ComplexityWalker extends tslint.RuleWalker {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar use `SyntaxWalker` instead

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+  private complexity = 0;
+  private nesting = 0;
+  private logicalOperationsToIgnore: ts.Node[] = [];
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar can we use a more precise type here?

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+      if (is(node.elseStatement, ts.SyntaxKind.IfStatement)) {
+        this.walk(node.elseStatement);
+      } else {
+        this.addComplexityWithoutNesting();
+        this.visitWithNesting(node.elseStatement);
+      }
+    }
+  }
+
+  private isEleseIf(node: ts.IfStatement): boolean {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar type in name

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+    return (
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar what for this parentheses?

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+  public visitConditionalExpression(node: ts.ConditionalExpression): void {
+    this.addComplexityWithNesting();
+    this.walk(node.condition);
+    this.visitWithNesting(node.whenTrue);
+    this.visitWithNesting(node.whenFalse);
+  }
+
+  public visitBinaryExpression(node: ts.BinaryExpression): void {
+    const opertator = node.operatorToken;
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar typo

File: sonarts-core/src/rules/cognitiveComplexityRule.ts [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+  }
+
+  private static isOperator(node: ts.Node, operatorKind: ts.BinaryOperator): boolean {
+    if (is(node, ts.SyntaxKind.BinaryExpression)) {
+      return is((node as ts.BinaryExpression).operatorToken, operatorKind);
+    }
+    return false;
+  }
+
+  private static skipParentheses(node: ts.Node): ts.Node {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/cognitiveComplexityRule.ts]: @m-g-sonar there is `drillDownThroughParenthesis` in navigation

File: sonarts-core/src/rules/cognitiveComplexityRule.ts (hunk 9 above)
Comment (m-g-sonar) [reply sonarts-core/src/rules/cognitiveComplexityRule.ts]: Yup, we only add `ts.BinaryExpression` here.

File: sonarts-core/src/rules/cognitiveComplexityRule.ts (hunk 10 above)
Comment (m-g-sonar) [reply sonarts-core/src/rules/cognitiveComplexityRule.ts]: I think it was sounding nice. but I'll fix :)

File: sonarts-core/src/rules/cognitiveComplexityRule.ts (hunk 11 above)
Comment (m-g-sonar) [reply sonarts-core/src/rules/cognitiveComplexityRule.ts]: That's not from me. It's Prettier which add it.

**Pull request merged**